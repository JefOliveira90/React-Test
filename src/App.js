import React, {Component} from 'react';
import {View} from 'react-native';

import commonStyles from './styles/common.js';

import Router from './components/Router.js';
import TitleBar from './components/TitleBar.js';

import Search from './views/Search.js';
import Historic from './views/Historic.js';
import SearchResults from './views/SearchResults.js';

export default class App extends Component {
    constructor(props) {
        super(props);

        this.state = {
            loading: false,
            search: '',
            searchError: null,
            searchResults: [],
            searchPage: 1,
            searchResultsPerPage: 10,
            searchTotalResults: 0,
            searchHistory: [],
            viewHistory: []
        };
    }

    async fetchSearchResults() {
        this.state.loading = true;

        try {
            let term = this.state.search;
            let page = this.state.searchPage;
            let url = `http://www.omdbapi.com/?apikey=2b1a6d43&s=${term}&page=${page}`;
            let request = await fetch(url);
            let response = await request.json();

            if(response.Response === 'True') {
                let results = response.Search;

                for(let i in results) {
                    let result = results[i];
                    let details = await this.fetchDetailedInfo(result.imdbID);

                    result = Object.assign(result, details);
                }

                this.setState((state, props) => ({
                    searchTotalResults: response.totalResults,
                    searchResults: state.searchResults.concat(results)
                }));
            } else {
                this.setState(() => ({
                    searchError: response.Error
                }));
            }
        } catch(reason) {
            console.error(reason);
        }

        this.state.loading = false;
    }

    async fetchDetailedInfo(imdbID) {
        let obj = {};

        try {
            let url = `http://www.omdbapi.com/?apikey=2b1a6d43&i=${imdbID}&plot=full`;
            let request = await fetch(url);

            obj = await request.json();
        } catch(reason) {
            console.error(reason);
        }

        return obj;
    }

    fetchMoreResults() {
        let limit = this.state.searchResultsPerPage;
        let total = this.state.searchTotalResults;

        if(this.state.loading || this.state.searchPage >= Math.ceil(total / limit)) {
            return;
        }

        this.setState((state) => ({
            searchPage: state.searchPage + 1
        }), this.fetchSearchResults);
    }

    resetSearchResults(callback) {
        this.setState((state, props) => {
            return {
                searchError: null,
                searchResults: [],
                searchPage: 1
            }
        }, callback);
    }

    setSearchTerm(term) {
        this.setState((state, props) => ({ search: term }));
    }

    registerSearchHistory(term) {
        let date = new Date();

        this.setState((state, props) => {
            state.searchHistory.push({
                key: term + date.toISOString(),
                term: term,
                date: date.toLocaleDateString()
            });
            
            return state;
        });
    }

    render() {
        return (
            <Router 
                app={this}
                initialView='search'
                views={{
                    'search': {
                        type: Search
                    },
                    'search_results': {
                        type: SearchResults
                    },
                    'historic': {
                        type: Historic
                    }
                }}/>
        );
    }
}