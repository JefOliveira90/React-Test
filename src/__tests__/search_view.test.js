import React from 'react';
import renderer from 'react-test-renderer';
import "isomorphic-fetch";

import App from '../App.js';

jest.mock('NativeAnimatedHelper');
jest.useFakeTimers();

jest.mock('TouchableNativeFeedback', () => {
    const ReactNative = require('react-native');

    return ReactNative.TouchableHighlight;
}) ;

test('Search view looks correct', () => {
    const app = renderer.create(<App />);
    const appInstance = app.getInstance();

    appInstance.setSearchTerm('Godfather');

    expect(app.toJSON()).toMatchSnapshot();
});