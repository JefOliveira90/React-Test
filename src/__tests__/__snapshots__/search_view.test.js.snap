// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search view looks correct 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#411a9b",
        "display": "flex",
        "flex": 10,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flex": 10,
          "justifyContent": "center",
          "width": "100%",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontSize": 18,
            "marginBottom": 20,
            "textAlign": "center",
            "width": "80%",
          }
        }
      >
        Insira o nome do filme ou s√©rie.
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Nome"
        placeholderTextColor="rgba(255, 255, 255, 0.4)"
        selectionColor="rgba(255, 255, 255, 0.4)"
        style={
          Object {
            "borderBottomColor": "white",
            "borderBottomWidth": 2,
            "color": "white",
            "fontSize": 28,
            "textAlign": "center",
            "width": "70%",
          }
        }
        underlineColorAndroid="transparent"
        value="Godfather"
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-around",
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "opacity": 0,
            "transform": Array [
              Object {
                "translateY": 40,
              },
            ],
            "width": 100,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "justifyContent": "center",
                "padding": 10,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../assets/icon-search.png",
                }
              }
              style={
                Object {
                  "height": 26,
                  "width": 26,
                }
              }
            />
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "white",
                }
              }
            >
              Buscar
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
